core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala

```diff
@@ -0,0 +1,425 @@
+
+
+class RequestQuotaTest extends BaseRequestTest {
+
+  override def numBrokers: Int = 1
+
+  private val topic = "topic-1"
+  private val numPartitions = 1
+  private val tp = new TopicPartition(topic, 0)
+  private val unthrottledClientId = "unthrottled-client"
+  private val brokerId: Integer = 0
+  private var leaderNode: KafkaServer = null
+
+  // Run tests concurrently since a throttle could be up to 1 second because quota percentage allocated is very low
+  case class Task(val apiKey: ApiKeys, val future: Future[_])
+  private val executor = Executors.newCachedThreadPool
+  private val tasks = new ListBuffer[Task]
+
+  override def propertyOverrides(properties: Properties): Unit = {
+    properties.put(KafkaConfig.ControlledShutdownEnableProp, "false")
+    properties.put(KafkaConfig.OffsetsTopicReplicationFactorProp, "1")
+    properties.put(KafkaConfig.OffsetsTopicPartitionsProp, "1")
+    properties.put(KafkaConfig.AuthorizerClassNameProp, classOf[RequestQuotaTest.TestAuthorizer].getName)
+    properties.put(KafkaConfig.PrincipalBuilderClassProp, classOf[RequestQuotaTest.TestPrincipalBuilder].getName)
+  }
+
+  @Before
+  override def setUp() {
+
+    RequestQuotaTest.principal = KafkaPrincipal.ANONYMOUS
+    super.setUp()
+
+    TestUtils.createTopic(zkUtils, topic, numPartitions, 1, servers)
+    leaderNode = servers.head
+
+    // Change default client-id request quota to a small value and a single unthrottledClient with a large quota
+    val quotaProps = new Properties()
+    quotaProps.put(DynamicConfig.Client.RequestPercentageOverrideProp, "0.01")
+    AdminUtils.changeClientIdConfig(zkUtils, "<default>", quotaProps)
+    quotaProps.put(DynamicConfig.Client.RequestPercentageOverrideProp, "2000")
+    AdminUtils.changeClientIdConfig(zkUtils, unthrottledClientId, quotaProps)
+
+    TestUtils.retry(10000) {
+      val quotaManager = servers(0).apis.quotas.request
+      assertEquals(s"Default request quota not set", Quota.upperBound(0.01), quotaManager.quota("some-user", "some-client"))
+      assertEquals(s"Request quota override not set", Quota.upperBound(2000), quotaManager.quota("some-user", unthrottledClientId))
+    }
+  }
+
+  @After
+  override def tearDown() {
+    try {
+      executor.shutdownNow()
+    } finally {
+      super.tearDown()
+    }
+  }
+
+  @Test
+  def testResponseThrottleTime() {
+
+    for (apiKey <- RequestQuotaTest.ClientActions) {
+      val builder = requestBuilder(apiKey)
+      submitTest(apiKey, () => {
+        checkRequestThrottleTime(apiKey)
+      })
+    }
+    waitAndCheckResults()
+  }
+
+  @Test
+  def testUnthrottledClient() {
+
+    val socket = connect()
+    try {
+      var correlationId: Int = 1
+      for (apiKey <- RequestQuotaTest.ClientActions) {
+        correlationId += 1
+        val response = requestResponse(socket, unthrottledClientId, correlationId, requestBuilder(apiKey))
+      }
+    } finally {
+      socket.close()
+    }
+    waitAndCheckResults()
```

Comment:
Hmm, is the test added? I don't see the code for submitTest that checks the throttling field in the response.